prov/util/src/util_mem_monitor.c

```diff
@@ -265,6 +266,20 @@ void ofi_monitor_notify(struct ofi_mem_monitor *monitor,
 	}
 }
 
+/* Must be called holding monitor lock */
+void ofi_monitor_flush(struct ofi_mem_monitor *monitor)
+{
+	struct ofi_mr_cache *cache;
+
+	dlist_foreach_container(&monitor->list, struct ofi_mr_cache,
```

Comment:
This list iteration isn't thread safe.  So the locking in this function doesn't provide protection...

The lock is being released because ofi_mr_cache_flush() acquires mm_lock...

I think we need 2 versions of ofi_mr_cache_flush() -- one which acquires the lock and another that assumes it's held (or pass in a bool need/have_lock).  Having this would let us simplify the locking in ofi_mr_cache_search(), which releases/acquires the mm_lock similar to what's done here, and for the same reasoning.

Reply:
Hmm, possibly, but I don't think that will resolve this issue. I think we need a separate lock for this dlist since ofi_mr_cache_flush must release the lock for its free anyway

Reply:
I switched to using the ts version but I'm not convinced there will not be a deadlock, thinking about it more.

Reply:
If the workflow goes:

1. Monitor is triggered from free
2. dlist_ts acquires lock
3. Cache flush is called
4. Cache hits a free event
5. Monitor is triggered from free
6. dlist_ts waits on a lock that won't be released

Sounds like a deadlock to me.

Now that I'm thinking about it more I don't see how we can have this dlist at all if we need to free inside of it.

Reply:
I agree -- I think there's a deadlock here.

Reply:
Since the lock in cache flush must be released in order to prevent another deadlock when calling cache flush itself, the dlist can never be protected. The previous implementation inside of ofi_mr_cache_notify had the same issue but it was less clear. As it stands we either can't have a dlist or can't free inside of it. The best I can think of right now is a large array of pointers to caches, but it's not scalable.

Reply:
I guess one question is how likely are we to hit this race condition?  Even with an array, we could remove a cache while walking it.

There may be some asynchronous method we could use to add/del caches from a monitor, but I'm not sure if it would be simple or even work.

If we ignore the list problem, with these changes, we're looking at possible recursion.  I can't tell if that's a real problem yet.

Reply:
It shouldn't be a problem because it's limited to number of entries on the list (or in the tree). If there's no entries (ie. the free is hooked into but not for a region we care about), it'll just be a no-op and then it will clear the stack.

Reply:
I'm thinking we can do something like this even if it's a bit janky:
```
modify is a global extern int
walk is a global extern int
monitor->new_lock will be a new lock created just for this dlist
This only works if add cache and del cache functions do not free in the section specified
(I looked at the code it just removes or adds to list, no free is done).

add cache function
-------------------
lock = monitor->new_lock;
before:
lock(lock)
if (modify || walk) {
	unlock(lock);
	goto before;
} else {
	modify = 1;
	unlock(lock);
}
dlist_insert
lock(lock);
modify = 0;
unlock(lock);

del cache function
--------------------
lock = monitor->new_lock;
before:
lock(lock)
if (modify || walk) {
	unlock(lock);
	goto before;
} else {
	modify = 1;
	unlock(lock);
}
dlist_remove
lock(lock);
modify = 0;
unlock(lock);

before monitor notify or flush function
--------------------
lock = monitor->new_lock;
before:
lock(lock);
if (modify) {
	unlock(lock);
	goto before;
} else {
	walk++;
	unlock(lock);
}
dlist loop - monitor notify/flush
lock(lock);
walk--;
unlock(lock);
```

Reply:
This is essentially a reader-writer locking mechanism, implemented by spinning.  At first glance, I think a reader/writer lock would work.  And as long as a waiting writer doesn't block a new reader, the reader lock should be safe to acquire recursively.

We don't have an abstraction for a reader-writer lock already.  And I don't believe we can use the pthread implementation for this because we need new readers to have priority over waiting writers

Reply:
If you're okay with it, I can implement the way I described before taking some more time to write a good reader/writer locking abstraction for libfabric (Needs more thinking to optimize and have configurable to give different priorities.)

Reply:
I would still pull this into a simple abstraction (ofi_rdpri_rwlock?), even if the implementation is basically as shown. If possible (we're not holding any spinlocks already), I'd swap in mutexes where you have locks, which should help with hard spinning.  We would need to document that the implementation is intentionally unfair to writers, and could starve them if used in a more generic code flow.

I think the code will be clearer if someone sees ofi_rdpri_wlock() / ofi_rdpri_wunlock() etc., rather than an open coded lock.

I also think ofi_rdpri_rwlock is a lousy name, but it tries to capture how the lock works... ofi_unfair_rwlock?  ofi_starving_rwlock?  ofi_mostly_read_rwlock?  Hmm... maybe unfair, not sure how easily someone gets 'reader prioritized' out of rdpri.

Reply:
Looking over pthread_rwlock, it looks like we may be able to loop on a nonblocking read-write lock

Reply:
Alright I'll go with ofi_rdpri_rwlock and if we want to change the name later I suppose we can

Reply:
Ah, yes, I think we may be able to use pthread_rwlock_rdlock combined with pthread_rwlock_trywrlock being used in a loop.  That would avoid needing to add an abstraction, and give us a working/optimized implementation that doesn't involve a roll-your-own lock.

Reply:
With pthread, we can check for EDEADLK to handle the case where a thread ends up recursively calling through.

Reply:
I'm not sure where EDEADLK would come in place, I updated with pthread_rwlock changes though.

Reply:
EDEADLK will only happen if a thread tries to acquire the write lock when it already holds it. That won't happen unless we try to add/remove a cache while adding/removing a cache.

Obtaining the read lock recursively is fine as long as the number of unlocks match the number of locks in the thread.